lib/AnalysisStructured/PtrAnalysis.cpp

```diff
@@ -0,0 +1,1128 @@
+  }
+
+  auto dstShape =
+      expandDimsOp.getResult().getType().cast<ShapedType>().getShape();
+  auto axis = expandDimsOp.getAxis();
+
+  assert(dstShape[axis] == 1 &&
+         "expect changed dimension to be 1 in expand_dims");
+
+  // insert dimension info
+  state.offsets.insert(state.offsets.begin() + axis, builder.getIndexAttr(0));
+  state.sizes.insert(state.sizes.begin() + axis, builder.getIndexAttr(1));
+  state.strides.insert(state.strides.begin() + axis, builder.getIndexAttr(0));
+  state.modulos.insert(state.modulos.begin() + axis, builder.getIndexAttr(0));
+
+  if (state.hasModulo() && state.getRank() > 2) {
+    expandDimsOp->emitRemark(
+        "PtrAnalysis: unsupported scenario where expand_dims result "
+        "has modulo and rank > 2");
+    return failure();
+  }
+
+  return success();
+}
+
+LogicalResult
+PtrAnalysis::visitOperandBroadcast(triton::BroadcastOp broadcastOp,
+                                   PtrSState &state, const Location loc,
+                                   OpBuilder &builder) {
+  assert(state.isEmpty());
+
+  auto src = broadcastOp.getSrc();
+  auto dst = broadcastOp.getResult();
+
+  SmallVector<int64_t> srcShape;
+  auto dstShape = dst.getType().cast<ShapedType>().getShape();
+
+  if (src.getType().isa<ShapedType>()) {
+    srcShape =
+        SmallVector<int64_t>(src.getType().cast<ShapedType>().getShape());
+    assert(srcShape.size() == dstShape.size() &&
+           "rank of source and destination should match");
+
+    if (visitOperand(src, state, loc, builder).failed()) {
+      return failure();
+    }
+
+    for (size_t i = 0; i < dstShape.size(); i++) {
+      if (srcShape[i] == dstShape[i]) {
+        continue;
+      } else if (srcShape[i] < dstShape[i]) {
+        state.sizes[i] = builder.getIndexAttr(dstShape[i]);
+      } else {
+        llvm_unreachable("unexpected dimensions used in broadcast");
+      }
+    }
+    return success();
+  }
+
+  broadcastOp->emitRemark("PtrAnalysis: Unsupported broadcast source type");
+  return failure();
+}
+
+LogicalResult PtrAnalysis::visitOperandSplat(triton::SplatOp splatOp,
+                                             PtrSState &state,
+                                             const Location loc,
+                                             OpBuilder &builder) {
+  assert(state.isEmpty());
+
+  auto src = splatOp.getSrc();
+  auto dst = splatOp.getResult();
+  auto dstShape = dst.getType().cast<ShapedType>().getShape();
+
+  if (visitOperand(src, state, loc, builder).failed()) {
+    return failure();
+  }
+
+  if (src.getType().isa<IntegerType, IndexType, triton::PointerType>()) {
+    for (auto s : dstShape) {
+      state.offsets.push_back(builder.getIndexAttr(0));
+      state.sizes.push_back(builder.getIndexAttr(s));
+      state.strides.push_back(builder.getIndexAttr(0));
+      state.modulos.push_back(builder.getIndexAttr(0));
+    }
+  } else {
+    splatOp->emitRemark("PtrAnalysis: unsupported splat pattern");
```

Comment:
The original code in the else clause is the following, which is admittedly complicated:

```cpp
    // src is a memref that represent a scalar pointer; it should have
    // one dimension of size 1. This happens inside a for loop that
    // originally has an init arg that is a tensor of pointers; this arg
    // would have been replaced by rewriteForOp.
    auto srcType = src.getType().cast<MemRefType>();
    assert(srcType.getRank() == 1 && state.getRank() == 1 &&
           "splat MemRef source should have rank 1");
    assert(srcType.getShape()[0] == 1 &&
           getIntAttr(state.sizes[0]).value() == 1 &&
           "splat MemRef source should have size 1");

    // Stride[0] will have value of 1 set in visitOperandAddPtr. This
    // value will be represented by a constOp. Clear this value.
    state.strides[0] = rewriter.getIndexAttr(0);

    for (auto [i, s] : llvm::enumerate(dstShape)) {
      if (i == 0) {
        state.sizes[i] = rewriter.getIndexAttr(s);
        continue;
      }
      state.offsets.push_back(rewriter.getIndexAttr(0));
      state.sizes.push_back(rewriter.getIndexAttr(s));
      state.strides.push_back(rewriter.getIndexAttr(0));
      state.modulos.push_back(std::nullopt);
    }
```

Is this something we won't support anymore?